include/Dialect/Poly/IR/PolyTypes.td

```diff
@@ -2,14 +2,11 @@
 #define HEIR_INCLUDE_DIALECT_POLY_IR_POLYTYPES_TD_
 
 include "PolyDialect.td"
+include "PolyAttributes.td"
 
 include "mlir/IR/DialectBase.td"
 include "mlir/IR/AttrTypeBase.td"
 
-//===----------------------------------------------------------------------===//
-// Poly type definitions
```

Comment:
It was mostly to keep the PR small enough. I think now that I'm deep in the dialect conversion framework, I see that, at the very least, we need the size of the underlying storage type and the coefficient bit width available in the type in order to use the TypeConverter.

Reply:
Mh, I'm not sure we want details of the TypeStorage to leak into the IR, rather the coefficient bit width should determine (in some implementation defined way) how the TypeStorage stores this.  My motivation for this is to have a clear separation between "MLIR the concept" and "MLIR the C++ code base", making sure the things we standardized depend only on the former (of course, with a reference implementation using the latter).

In order to make the TypeConverter work, you could add a `storageType()` method to the (C++ implementation of) the type. Alternatively, the type could just figure this out internally in it's `create`/`build`/constructor, especially if we end up using something like `MPint` anyway?

Reply:
I see! I guess in an ideal world we'd want some kind of `Z_q` type for the second part (i.e., instead of `i32`)? Sort of a `baseQ` (taking inspiration from the `base2` dialect proposed by @KFAFSP)

Reply:
I think the polynomial attribute in this PR would be sufficient to put on the type. And then we just mandate that each polynomial type is associated with a specific ring, and if we want to switch rings we have to add some special ops for that.

Reply:
For example, `base2.fixed<signed 32,16>` (alias `base2.si32_16`) is only usable with a compile-time fixed precision. Is this desired here, or do you need a type to model $\mathbb{Q}$ instead of $\mathbb{Z}_N \cdot 2^{-E}$?

Reply:
> I think the polynomial attribute in this PR would be sufficient to put on the type. And then we just mandate that each polynomial type is associated with a specific ring, and if we want to switch rings we have to add some special ops for that.

I'm not sure I'm following. In my understanding, an attribute is a compile-time value (usually closely associated with a type, with a `dialect.constant` operation that "turns" an attribute into a value of that type). Having a `#poly.polynomial` attribute as a type parameter for the `!poly.polynomial` type would be parameterizing the type by a specific polynomial? Sorry if I'm misunderstanding things here!  I guess we could parameterize the `!poly.polynomial` type by `#poly.ring` if that's what you meant?